google-cloud-testing/google-cloud-bigtable-emulator/README.md

```diff
@@ -0,0 +1,181 @@
+testCompile 'com.google.cloud:google-cloud-bigtable-emulator:0.69.0-alpha'
+testCompile 'junit:junit:4.12'
+```
+If you are using SBT, add this to your dependencies
+```Scala
+libraryDependencies += "com.google.cloud" % "google-cloud-bigtable" % "0.69.0-alpha"
+libraryDependencies += "com.google.cloud" % "google-cloud-bigtable-admin" % "0.69.0-alpha"
+libraryDependencies += "com.google.cloud" % "google-cloud-bigtable-emulator" % "0.69.0-alpha" % Test
+libraryDependencies += "junit" % "junit" % "4.12" % Test
+```
+[//]: # ({x-version-update-end})
+
+## Getting Started
+
+Here is a code snippet showing a simple JUnit test. Add the following imports
+at the top of your file:
+
+```java
+import com.google.api.core.ApiFuture;
+import com.google.api.gax.core.NoCredentialsProvider;
+import com.google.api.gax.grpc.GrpcTransportChannel;
+import com.google.api.gax.rpc.FixedTransportChannelProvider;
+import com.google.cloud.bigtable.admin.v2.BigtableTableAdminClient;
+import com.google.cloud.bigtable.admin.v2.BigtableTableAdminSettings;
+import com.google.cloud.bigtable.admin.v2.models.CreateTableRequest;
+import com.google.cloud.bigtable.data.v2.BigtableDataClient;
+import com.google.cloud.bigtable.data.v2.BigtableDataSettings;
+import com.google.cloud.bigtable.data.v2.models.Row;
+import com.google.cloud.bigtable.data.v2.models.RowMutation;
+import com.google.cloud.bigtable.emulator.v2.BigtableEmulatorRule;
+import java.io.IOException;
+import java.util.concurrent.ExecutionException;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+```
+
+Then, to make a query to Bigtable, use the following code:
+```java
+@RunWith(JUnit4.class)
+public class ExampleTest {
+  // Initialize the emulator Rule
+  @Rule
+  public final BigtableEmulatorRule bigtableEmulator = BigtableEmulatorRule.create();
+
+  // Clients that will be connected to the emulator
+  private BigtableTableAdminClient tableAdminClient;
+  private BigtableDataClient dataClient;
+
+  @Before
+  public void setUp() throws IOException {
+    // Initialize the clients to connect to the emulator
+    BigtableTableAdminSettings.Builder tableAdminSettings = BigtableTableAdminSettings.newBuilder()
+        .setInstanceName(com.google.bigtable.admin.v2.InstanceName.of("fake-project", "fake-instance"));
+    tableAdminSettings.stubSettings()
+        .setCredentialsProvider(NoCredentialsProvider.create())
+        .setTransportChannelProvider(
+            FixedTransportChannelProvider.create(
+                GrpcTransportChannel.create(bigtableEmulator.getAdminChannel())
+            )
+        );
+    tableAdminClient = BigtableTableAdminClient.create(tableAdminSettings.build());
+
+    BigtableDataSettings.Builder dataSettings = BigtableDataSettings.newBuilder()
+        .setInstanceName(com.google.cloud.bigtable.data.v2.models.InstanceName.of("fake-project", "fake-instance"))
```

Comment:
I wonder if this can be added as default way to get a data client to `BigtableEmulatorRule`, something like `getBigtableDataClient`. The same for the table admin client above.

Reply:
I would like to use this wrapper for tests in the bigtable client, which would create a circular dependency: bigtable client depends on the emulator for tests and the emulator depends on the client to construct it.
I'm considering adding the emulator as an optional dependency to the client and then adding a helper to construct the clients for the emulator. But I need to do a bit more experimentation.